2주차 졸업과제: movieflix 제출

DaehunGwak opened the pull request:
안녕하세요! 이런식으로 제출하고 리뷰과정을 진행하면 될지 확인 부탁드립니다 :)

### 과제 공유 링크
https://daehun.notion.site/12-Movieflix-6b70adf36c8d4976a6a82db06ed17c17?pvs=4

### 데모 링크
https://imgur.com/a/wznzb4G

### 깨달은 점
<details>

- Hero tag 가 계속 중복하는 오류가 발생 > movie id 가지고만 Hero tag 를 만드니 홈 스크린에서 여러 영화가 중복해서 나와 tag duplicated 현상이 발생하여 각 영역마다 tag를 달리 가질 수 있도록 개선
- detail view 로 넘어갈 시 화질 깨짐 발생 > tmdb 이미지 규격들을 살펴보니 w500 규격만 있는게 아니라 더 고화질인 w1280 규격을 서빙하고 있음을 발견, 이를 적용
- w1280 사용 시 로딩 시간이 조금 발생 >  `precacheImage` 를 활용해 각 카드 리스트 빌딩 시 고화질 이미지도 같이 캐싱되도록 구현, detail screen 으로 넘어갈 시 이미지 로딩을 느끼지 못하도록 구현
    - precacheImage 메소드도 비동기 이벤트루프 방식이므로 동기식에서 동작 시, 별도의 백그라운드 프로세스 스케줄러에서 동작하여 따로 메인 ui 플로우에 큰영향을 주지 않음 (물론 헤비한 동작이면 Hardware 에 영향을 주어 앱 전체에 영향 가능)
    - 자세한 사항은 https://dart.dev/language/concurrency 링크 참고
- 별점 구현 시, 별을 부분적으로 칠하고 싶다는 개인적인 욕심이 생김 > ShaderMask 를 사용하여 구현
    - 코드 참고: https://github.com/DaehunGwak/flutter-playground/blob/d943a90e7fa9b9acc95103bf1071cd478faf3c46/movieflix/lib/widgets/movie_datail_widget.dart#L148-L172
- detail api 값들을 보니 popularity 와 vote_avarage 필드 값의 역할이 처음에 헷갈림 > populairty 는 아마 `/popular` api 을 구현할 때 사용하는 스코어로 보여지고, vote_avarage 가 실제 유저가 투표한 값이라는 의도가 더 보이는 것 같아 vote_avarage 사용 결정
- ApiService 처럼 실제로 인스턴스화 할 필요 없는 클래스를 만들 시, 따로 인스턴스화 하는 실수를 방지하기 위해 constructor를 `ApiService._();` 로 지정하면 인스턴스화가 불가능
</details>

### 개선할 점
<details>

- Hero animation 시 배경 오른쪽에 잠시 블랙 배경이 들어나는게 보여 개선 필요
- font 깨짐 현상 발생, 다국어 지원 폰트를 찾아 적용하는 것이 필요
- 별점 구현 시, 요구된 이미지에서는 배경색이 opacity 가 적용되어 있어 적용 필요
- 각 위젯 클래스의 책임이 너무 커진 곳들이 있음 > 별도의 클래스로 나누는 리팩터링 필요
</details>

DaehunGwak committed:
movieflix: 프로젝트 초기화

DaehunGwak committed:
movieflix: 프로젝트 위젯 및 로직 추가

devgony commented on movieflix/analysis_options.yaml:
```diff
@@ -0,0 +1,28 @@
+  # `// ignore_for_file: name_of_lint` syntax on the line or in the file
+  # producing the lint.
+  rules:
+    # avoid_print: false  # Uncomment to disable the `avoid_print` rule
```
아래 내용을 추가 하신 뒤 project root 에서 `dart fix --apply` 를 수행하시면 대부분의 
 경우에 `,` 를 붙여서 좀 더 깔끔한 formatting 을 하실 수 있습니다. 
해당내용은 CI 에도 걸어서 다들 일관된 옵션을 가지시도록 해볼까 고려 중입니다.
```dart
// analysis_options.yaml
linter:
  rules:
    require_trailing_commas: true
```

DaehunGwak replied:
@devgony 님 감사합니다! 엄청 편하네요~
다음엔 IDE 에 자동으로 반영 시켜놓겠습니다.
관련 커밋: https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/72493ebedb0ca5bb9002bfd00087f02639d707a5

devgony replied:
엇.. 혹시 analysis_options.yaml 에 반영을 해보실 수 있을까요? 

현재는 pubspec.yaml 에 반영이 된 것 같습니다. ㅠ

DaehunGwak replied:
@devgony 님,
제가 착각하고 잘못 반영했네요 ㅎㅎ 아래 커밋에 다시 반영해두었습니다!
관련 커밋: https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/87ce421f1f50c76f4ac8b60b55ccc5383010fe10

DaehunGwak committed:
movieflix: linter 적용 및 .gitignore 수정

DaehunGwak committed:
movieflix: linter 적용 추가 수정

DaehunGwak committed:
movieflix: ide 자동 반영 및 전체 파일 포매팅

devgony commented on movieflix/lib/widgets/movie_card_widget.dart:
```diff
@@ -0,0 +1,198 @@
+          height: isDisplayTitle ? (imageHeight + 80) : (imageHeight + 10),
+          child: FutureBuilder(
+            future: movieApiFuture,
+            builder: (context, snapshot) {
```
FutureBuilder 에서 error 가 있는 경우에 대한 처리가 되면 더 좋을 것 같아요!
에러 페이지로 이동시키는 법도 있겠지만 `SnackBar` 를 통해 하단에 메시지를 띄울 수도 있습니다.

```dart
builder: (context, snapshot) {
  if (snapshot.hasError) {
    WidgetsBinding.instance.addPostFrameCallback((_) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('서버 점검중입니다. 관리자에게 문의해주세요')),
      );
    });
    return const Center(child: CircularProgressIndicator());
  }
..
```

DaehunGwak replied:
@devgony 님 감사합니다!
덕분에 `SnackBar`, `WidgetsBinding.instance.addPostFrameCallback`, `ScaffoldMessenger` 를 알게되었습니다.
관련 코드는 오른쪽 커밋에서 확인 부탁드리겠습니다 :) (commit: https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/769f59a28a11413e2247139493f5a48b42a2dddb)

![image](https://github.com/nlutter-study/nlutter-study-2nd/assets/12469427/e181f70d-df96-421a-81e8-cb27a98cd38c)

devgony commented on movieflix/lib/services/api_service.dart:
```diff
@@ -0,0 +1,70 @@
+  static Future<List<SimpleMovie>> getPopularMovies() async {
+    final uri = Uri.parse('$_baseUrl/$_popularPath');
+    final response = await http.get(uri);
+    _handleErrorStatusCode(response);
```
`http.get` 이 Exception을 throw 해서 `_handleErrorStatusCode` 에는 실제로는 도달하지 않는 것으로 보입니다. 커스텀 에러메시지를 throw 하기를 원하시는 경우 try catch 로 감쌀 필요가 있을거같아요!

DaehunGwak replied:
@devgony 님 감사합니다! 일단 무조건 응답 오는 줄 알았네요 ㅠ
해당 부분 일단 기초적인 예외만 잡아서 커스텀 로그 남기는 방식으로 가져가 보았습니다.
(관련 커밋: https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/a9a3c8b599edec61faaf875bb24bf9c6af22d513)

추가적으로 404 에러가 나도록 path 를 조정하여 테스트 해보니,
따로 예외를 뱉지 않고 그대로 요청 아래 로직으로 넘어가는 현상이 있어 어떻게 좋은 핸들링을 할지는 더 고민해보도록 하겠습니다.

#### 관련 http 요청 내부 코드
![image](https://github.com/nlutter-study/nlutter-study-2nd/assets/12469427/6ff56adb-22b0-4e40-98f3-8aefa34c323f)

#### ClientException 일부러 내서 테스트 후 로그 확인
![image](https://github.com/nlutter-study/nlutter-study-2nd/assets/12469427/a0b1d7d4-1419-4118-8183-480286fa7e17)

DaehunGwak replied:
reference: https://dart.dev/language/error-handling

devgony replied:
헛.. 이거는 제 실수네요 제가 테스트할 때 아예 url 패턴이 아닌걸 넣어서 바로 exception 이 나던거였습니다.
그래도 exception 이 날 가능성이 있다는 건 맞으니 try-catch 도 감싸주어야하고 statusCode 체크도 같이 해주어야겠네요

devgony commented on movieflix/lib/services/api_service.dart:
```diff
@@ -0,0 +1,70 @@
+    _printDebug(response.body);
+    final List<dynamic> results = _utf8JsonDecode(response.bodyBytes)[_resultsField];
+    return results.map((e) => SimpleMovie.fromJson(e)).toList();
+  }
```
단일 책임의 원칙에 의해 각 함수의 이름으로 지금처럼 각 api service 들을 세분화 할 수도있지만,
함수형 진영에서는 지금처럼 구현부가 거의 동일한 경우 함수명에만 존재하던 의미를 명시적인 parameter 로 위임하여 각 경우의 수를 first class value 로 관리하고 중복 코드도 줄이는 refactoring 을 하기도 합니다.

```suggestion
enum MovieType { simple, detail }
...
static Future<dynamic> fetchData(String path, MovieType movieType) // getMovieDetail 코멘트 참조
...
static Future<List<SimpleMovie>> getMovies(FetchType fetchType) async {
    final List<dynamic> data =
        await fetchData(fetchType.getUrl(_baseUrl), MovieType.simple);
    return data.map<SimpleMovie>((e) => SimpleMovie.fromJson(e)).toList();
 }
```

parameter인 fetchType 의 경우 exhaustive 한 패턴 처리를 위해 flutter 에서는 enum 또는 sealed class 를 활용할 수 있는데, 이번 리뷰의 목적 중 하나가 강의에는 없는 Dart 3 의 최신 기능들을 써보는 것이므로 sealed class 로 구현 해보았습니다.

```dart
sealed class FetchType {
  final String path;
  final String title;
  FetchType._(this.path, this.title);
  factory FetchType.popular() => Popular._("popular", "Popular Movies");
  factory FetchType.nowPlaying() =>
      NowPlaying._("now-playing", "Now in Cinemas");
  factory FetchType.comingSoon() => ComingSoon._("coming-soon", "Coming soon");

  getUrl(String baseUrl) => '$baseUrl/$path';
}

class Popular extends FetchType {
  Popular._(
    String path,
    String title,
  ) : super._(path, title);
}

class NowPlaying extends FetchType {
  NowPlaying._(
    String path,
    String title,
  ) : super._(path, title);
}

class ComingSoon extends FetchType {
  ComingSoon._(
    String path,
    String title,
  ) : super._(path, title);
}
```

DaehunGwak replied:
@devgony 님~ 덕분에 api 중복 코드 분량을 많이 줄일 수 있었습니다!
오른쪽 커밋에 반영해두었으니 한번 확인 부탁드릴께요! (https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/5f4c052dae03c8e8b20cbf0aaeb58f5dae8b390a)

devgony commented on movieflix/lib/services/api_service.dart:
```diff
@@ -0,0 +1,70 @@
+      throw Exception(
+          'API 요청 실패 (url: ${response.request?.url}, status code: ${response.statusCode}');
+    }
+  }
```
getMovies 와 getDetailMovie 가 데이터를 가져오는 부분이 거의 유사하여 fetchData 로 합쳐보았습니다. fetch 할 대상이 simple 인지 Detail인지는 enum MovieType 으로 묶었고 dart3 의 switch expression 을 사용하여 리턴해보았습니다.

또한 _handleErrorStatusCode이 unreachable 이던 점을 개선하기 위해 에러 핸들링 부분을 fetchData 안에 try-catch 로 합쳤습니다.

```dart
enum MovieType { simple, detail }
...
static Future<DetailMovie> getDetailMovie(int id) async {
  final data = await fetchData('$_baseUrl/$_detail$id', MovieType.detail);
  return DetailMovie.fromJson(data);
}
...
static Future<dynamic> fetchData(String path, MovieType movieType) async {
    try {
      final uri = Uri.parse(path);
      final response = await http.get(uri);
      final decoded = _utf8JsonDecode(response.bodyBytes);

      return switch (movieType) {
        MovieType.simple => decoded[_resultsField],
        MovieType.detail => decoded,
      };
    } catch (e) {
      debugPrint("Error: ${e.toString()}");
      throw Exception(
        'API 요청 실패',
      );
    }
  }
```

DaehunGwak replied:
오른쪽 커밋에 같이 반영해두었습니다! 리뷰 감사합니다 :) (https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/5f4c052dae03c8e8b20cbf0aaeb58f5dae8b390a)

devgony commented on movieflix/lib/services/api_service.dart:
```diff
@@ -0,0 +1,70 @@
+    if (kDebugMode) {
+      print(something);
+    }
+  }
```
debug 전용으로 print를 원하는경우 `debugPrint()` 를 사용하셔도 동일 효과를 낼 수 있고 라인 정렬도 해준다고합니다.

DaehunGwak replied:
해당 부분도 같이 반영해 두었습니다! (https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/a9a3c8b599edec61faaf875bb24bf9c6af22d513)

devgony commented on movieflix/lib/widgets/movie_card_widget.dart:
```diff
@@ -0,0 +1,198 @@
+import '../screens/detail_screen.dart';
+import '../utils/tag_builder.dart';
+
+class MovieCardWidget extends StatelessWidget {
```
이미지는 두 가지 종류의 사이즈만 가지므로 sealed class ImageType 을 선언하여 묶어서 받아 parameter의 개수를 줄였습니다.

fetchType 역시 parameter로 정의하고 movieApiFuture 와 title 을 묶어서 받게 되었습니다.

```dart
sealed class ImageType {
  final double height;
  final double width;
  ImageType._(this.height, this.width);
  factory ImageType.square() => Square._(150, 150);
  factory ImageType.landscape() => Landscape._(220, 320);
}

class Square extends ImageType {
  Square._(
    double height,
    double width,
  ) : super._(height, width);
}

class Landscape extends ImageType {
  Landscape._(
    double height,
    double width,
  ) : super._(height, width);
}

class MovieCardWidget extends StatelessWidget {
  const MovieCardWidget({
    super.key,
    required this.fetchType,
    required this.imageType,
    this.isDisplayTitle = true,
  });

  final FetchType fetchType;
  final ImageType imageType;
...
```

DaehunGwak replied:
@devgony 님, 오 sealed class 라는 것이 있군요!
imageType 에 대해서 반영해두었습니다! (https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/4ac38a297b25fd4d3b1510f3d714f4f9eef6134c)

DaehunGwak replied:
sealed class 관련 공부 레퍼런스: https://dart.dev/language/class-modifiers#sealed

DaehunGwak replied:
enum 과 sealed class 의 기능적 차이와 각각 어떨 때 쓸지 판단해볼 필요 존재

devgony commented on movieflix/lib/widgets/movie_card_widget.dart:
```diff
@@ -0,0 +1,198 @@
+          ),
+        ),
+        SizedBox(
+          height: isDisplayTitle ? (imageHeight + 80) : (imageHeight + 10),
```
ternary를 dart3 의 switch expression 으로 바꿔보았습니다.
```dart
height: imageType.height +
    switch (isDisplayTitle) {
      true => 80,
      false => 10,
    },
```

DaehunGwak replied:
오 switch expression 을 이런식으로 쓸 수 있군요 공유 감사합니다!
오른쪽 커밋에 반영해 두었습니다! (https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/28b69fcaad4481ac93dd4630007c076da8648db4)

devgony commented on movieflix/lib/widgets/movie_card_widget.dart:
```diff
@@ -0,0 +1,198 @@
+            future: movieApiFuture,
+            builder: (context, snapshot) {
+              if (!snapshot.hasData) {
+                return _EmptyCardListWidget(
+                  imageWidth: imageWidth,
+                  imageHeight: imageHeight,
+                  isDisplayTitle: isDisplayTitle,
+                );
+              }
+              return _CardListWidget(
+                imageWidth: imageWidth,
+                imageHeight: imageHeight,
```
두 위젯 모두 imageType 으로 묶어서 argument 를 넘겨 주었습니다.
```dart
return _EmptyCardListWidget(
  imageType: imageType,
  isDisplayTitle: isDisplayTitle,
);
}
return _CardListWidget(
imageType: imageType,
```

DaehunGwak replied:
imageType 한꺼번에 반영해두어서 같이 남깁니다~ (https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/4ac38a297b25fd4d3b1510f3d714f4f9eef6134c)

devgony commented on movieflix/lib/screens/home_screen.dart:
```diff
@@ -0,0 +1,52 @@
+          crossAxisAlignment: CrossAxisAlignment.start,
+          children: [
+            const SizedBox(height: 120),
+            MovieCardWidget(
+              movieApiFuture: ApiService.getPopularMovies(),
+              title: "Popular Movies",
+              imageHeight: 220,
+              imageWidth: 320,
+              isDisplayTitle: false,
+            ),
+            MovieCardWidget(
+              movieApiFuture: ApiService.getNowPlayingMovies(),
+              title: "Now in Cinemas",
+              imageHeight: 150,
+              imageWidth: 150,
+              isDisplayTitle: true,
+            ),
+            MovieCardWidget(
+              movieApiFuture: ApiService.getComingSoonMovies(),
+              title: "Coming soon",
+              imageHeight: 150,
+              imageWidth: 150,
+              isDisplayTitle: true,
+            ),
```
fetchType 과 ImageType 으로 argument 의 개수를 줄였습니다.
isDisplayTitle의 경우 default 값인 true 는 뺐습니다. 명시적으로 주는것도 좋다고 생각합니다.
```dart
MovieCardWidget(
  fetchType: FetchType.popular(),
  imageType: ImageType.landscape(),
  isDisplayTitle: false,
),
MovieCardWidget(
  fetchType: FetchType.nowPlaying(),
  imageType: ImageType.square(),
),
MovieCardWidget(
  fetchType: FetchType.comingSoon(),
  imageType: ImageType.square(),
```

DaehunGwak replied:
imageType 한꺼번에 반영해두어서 같이 남깁니다~ (https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/4ac38a297b25fd4d3b1510f3d714f4f9eef6134c)

DaehunGwak replied:
오른쪽 커밋에 fetchType 같이 반영해두었습니다! (https://github.com/nlutter-study/nlutter-study-2nd/pull/5/commits/5f4c052dae03c8e8b20cbf0aaeb58f5dae8b390a)

devgony reviewed (changes requested):
너무 깔끔하게 잘 해주셔서 리뷰 해드릴 거리를 찾느라 오래걸렸습니다ㅠ
리뷰 달아 드린 부분은 Dart3 신기능을 연습해보자 + 저의 개인 코딩스타일 이오니 다른 의견이나 더 나은 방안이 있으시면 언제든지 댓글로 달아주세요!

devgony commented:
CI 구성을 완료해서 각 {githubID} branch 에 넣어두었습니다.
아래처럼 fetch 후 merge해서 다시 올려주시면 액션이 수행됩니다!

```sh
git fetch review-upstream {githubID}
git merge review-upstream/{githubID}
git push review-origin {currentBranch}
```

DaehunGwak committed:
Merge remote-tracking branch 'review-upstream/DaehunGwak' into feature/movieflix

DaehunGwak committed:
fix: test 코드 trailing eof 추가

DaehunGwak committed:
fix: api 관련 FutureBuilder 에러 핸들링 수정

fix this review: https://github.com/nlutter-study/nlutter-study-2nd/pull/5#discussion_r1425317635

DaehunGwak committed:
fix: api 관련 예외처리 로직 수정

https://github.com/nlutter-study/nlutter-study-2nd/pull/5#discussion_r1426718488

DaehunGwak committed:
fix: tenary 연산을 dart3 switch expression 으로 수정

https://github.com/nlutter-study/nlutter-study-2nd/pull/5#discussion_r1426823726

DaehunGwak committed:
fix: sealed class 를 이용한 card image size type 화 수정

https://github.com/nlutter-study/nlutter-study-2nd/pull/5#discussion_r1426820663
https://github.com/nlutter-study/nlutter-study-2nd/pull/5#discussion_r1426827660
https://github.com/nlutter-study/nlutter-study-2nd/pull/5#discussion_r1426830498

DaehunGwak committed:
fix: sealed class, enum 를 이용한 ApiService 수정

https://github.com/nlutter-study/nlutter-study-2nd/pull/5#discussion_r1426778849
https://github.com/nlutter-study/nlutter-study-2nd/pull/5#discussion_r1426807108
https://github.com/nlutter-study/nlutter-study-2nd/pull/5#discussion_r1426830498

devgony commented on movieflix/lib/services/api_service.dart:
```diff
@@ -2,67 +2,84 @@ import 'dart:convert';
+
+      return switch (dataType) {
+        DataType.list => decoded[_resultsField],
+        DataType.object => decoded,
```
list 와 object 로 구분 해주신 것도 좋아보입니다 👍🏼

devgony reviewed (approved):
각 리뷰를 반영 해주실 때마다 관련 레퍼런스를 찾아주시고 스크린샷 까지 찍어주신 점 너무 멋집니다 👍🏼
고생 많으셨습니다!

DaehunGwak commented:
@devgony 님 리뷰감사합니다~!
제가 머지할 권한은 없어서 머지 해주신다면 감사하겠습니다!
